src/Week11/SExpr.hs

```diff
@@ -0,0 +1,54 @@
+module Week11.SExpr where
+
+import Control.Applicative
+import Data.Char
+import Week11.AParser
+
+------------------------------------------------------------
+--  1. Parsing repetitions
+------------------------------------------------------------
+
+zeroOrMore :: Parser a -> Parser [a]
+zeroOrMore p = oneOrMore p <|> pure []
+
+oneOrMore :: Parser a -> Parser [a]
+oneOrMore p = liftA2 (:) p $ zeroOrMore p
+
+------------------------------------------------------------
+--  2. Utilities
+------------------------------------------------------------
+
+spaces :: Parser String
+spaces = zeroOrMore $ satisfy isSpace
+
+ident :: Parser String
+ident = liftA2 (:) (satisfy isAlpha) (zeroOrMore $ satisfy isAlphaNum)
+
+------------------------------------------------------------
+--  3. Parsing S-expressions
+------------------------------------------------------------
+-- definitions
+type Ident = String
+
+data Atom = N Integer | I Ident
+  deriving Show
+
+data SExpr = A Atom
+           | Comb [SExpr]
+  deriving Show
+------------------------------------------------------------
+
+parseSExpr :: Parser SExpr
+parseSExpr = atomExpr <|> combExpr
+
+combExpr :: Parser SExpr
+combExpr = Comb <$> (open *> oneOrMore parseSExpr) <* close
+  where
+    open = char '(' *> spaces
+    close = char ')' <* spaces
+
+atomExpr :: Parser SExpr
+atomExpr = A <$> (atomIdent <|> atomInt) <* spaces
+  where
+    atomIdent =  I <$> ident
+    atomInt = N <$> posInt
```

Comment:
This might not parse the valid atom ` fooBar` because of the preceeding spaces.

Since the pattern of removing spaces before and after is common, you could also consider creating a combinator for that specific use case

```
withNoSpaces :: Parser a -> Parser a
```